File: app/domain/clustershell.go
package domain

import "errors"

// ErrShellNamespaceRequired is raised when an in-cluster shell is asked for
// without a namespace.
//
// IT EXISTS BECAUSE THE EMPTY STRING IS NOT AN ERROR ANYWHERE ELSE HERE.
// NewNamespaceName("") returns NamespaceAll, which every LIST in this
// application reads as "every namespace" — perfectly sensible for a read, and
// meaningless for a create: a pod lives in exactly one namespace, and handing
// the empty string to the client would create it in whatever the kubeconfig's
// default happens to be. That is a pod appearing somewhere nobody asked for,
// under a name that composed perfectly well. So "all namespaces" is refused
// and the operator is asked, rather than a namespace being guessed for them.
var ErrShellNamespaceRequired = errors.New("an in-cluster shell needs one namespace, and \"all namespaces\" is not one")

// ClusterShell is one live in-cluster shell — an ordinary pod PodSteer created
// in a namespace, running a shell its terminal session attaches to.
//
// THE THIRD THING PODSTEER CAN OPEN INSIDE A CLUSTER, and it is neither of
// the other two. It is not the ephemeral debug container, which is injected
// into somebody else's pod and shares that pod's namespaces; and it is not the
// node shell, which is privileged, enters a node's host namespaces with
// nsenter, and is the most powerful thing this application can do. This is an
// ordinary, unprivileged pod, admissible under Pod Security's `restricted`
// profile, for somebody who wants kubectl, dig and curl FROM INSIDE the
// cluster's network — a vantage point, not a privilege.
//
// Tracked exactly like NodeShell, and for the identical reason: the pod is a
// resource PodSteer created and is on the hook to remove, so the record of it
// and the thing that deletes it are created together and torn down together,
// the pod is deleted when the terminal session ends or PodSteer closes, and
// the live set is listable so the activity surface can show it with a stop
// control.
type ClusterShell struct {
	// ID is stable for the life of the shell.
	ID string
	// ClusterID, Namespace and PodName say where the pod is, so it can be
	// deleted against the cluster it was created on — the cluster is part of
	// the identity for the same reason it is on a NodeShell: two clusters run
	// identically named namespaces, and deleting the wrong one is worse than
	// leaking.
	ClusterID ClusterID
	Namespace NamespaceName
	PodName   string
	// Image is the pod's image, kept for display in the activity list.
	Image string
	// ContainerName is the pod's single container, which the terminal session
	// attaches to. Carried here so the layer opening the session need not know
	// how the pod was built.
	ContainerName string
	// Adopted reports that this record was taken over an EXISTING pod rather
	// than one this session created — the reuse path. It changes nothing about
	// the lifecycle (an adopted pod is deleted on session end exactly as a
	// created one is); it is here so the activity list and the log can say
	// which happened, because "PodSteer created a pod" and "PodSteer took
	// responsibility for one it found" are different sentences about a
	// cluster.
	Adopted bool
}

// ClusterShellCandidate is a pod PodSteer created for an in-cluster shell,
// found in a namespace before another one is created.
//
// PHASE IS QUOTED, NOT JUDGED. The candidate carries the pod's phase exactly
// as the API server reported it, and the decision about what may be offered is
// PlanClusterShellReuse's — see that function for why only a running pod is an
// offer.
type ClusterShellCandidate struct {
	PodName string
	Image   string
	// Phase is the pod's status.phase as the API server reported it —
	// PodPhaseRunning, PodPhasePending, PodPhaseSucceeded, PodPhaseFailed,
	// PodPhaseUnknown, or "" when the kubelet has not reported one yet.
	//
	// The PodPhase the pod list already uses rather than a plain string:
	// PlanClusterShellReuse compares it against PodPhaseRunning, and two
	// spellings of "Running" in one repository is one too many.
	// PodPhaseTerminating never appears here — that one is the mapper's own
	// substitution for a deleting pod, and this reads status.phase directly.
	Phase PodPhase
}

// ClusterShellReuse splits the pods PodSteer created in a namespace into the
// ones that may be offered and the ones that may only be reported.
type ClusterShellReuse struct {
	// Reusable are pods in a Running phase. Attaching to one of these lands
	// the operator on a shell.
	Reusable []ClusterShellCandidate
	// Other are pods PodSteer created that are in any other phase. They are
	// worth SAYING — a namespace accumulating them is the explanation for
	// pods somebody is looking at and cannot account for — but they are not
	// offers.
	Other []ClusterShellCandidate
}

// PlanClusterShellReuse decides which of PodSteer's own shell pods in a
// namespace may be offered for reuse.
//
// ONLY RUNNING, AND THAT IS THE WHOLE RULE. A pod that has exited — Succeeded
// because its shell was closed from inside, Failed because it hit its
// activeDeadlineSeconds backstop, Pending because it cannot pull its image —
// cannot be attached to. Offering one would hand somebody an attach that fails
// for a reason the offer gave them no way to see, which is worse than not
// offering: they asked for a shell and got an error about a pod they never
// chose to create.
//
// A pod in some other state is still REPORTED, in Other, because it answers a
// question the operator will otherwise have to ask the cluster: why does this
// namespace have five podsteer-shell-* pods in it. Reporting is not offering.
func PlanClusterShellReuse(candidates []ClusterShellCandidate) ClusterShellReuse {
	plan := ClusterShellReuse{}
	for _, candidate := range candidates {
		if candidate.Phase == PodPhaseRunning {
			plan.Reusable = append(plan.Reusable, candidate)
			continue
		}
		plan.Other = append(plan.Other, candidate)
	}
	return plan
}

File: web/src/lib/clusterShell.ts
/**
 * Pure helpers for the in-cluster shell dialog.
 *
 * Kept out of the component for the reason `debugShell.ts` is: the two rules
 * that matter here — which namespace, and which pods may be offered — are
 * decisions rather than rendering, and a decision belongs somewhere it can be
 * argued with in a test.
 */

import { ALL_NAMESPACES } from './api/client'
import { DEFAULT_CLUSTER_SHELL_IMAGE } from '$stores/preferences.svelte'

/** What a confirmed in-cluster shell dialog asks for. */
export interface ClusterShellRequest {
  image: string
  namespace: string
}

/**
 * The namespace the dialog opens on, given the tab's current filter.
 *
 * DEFAULTS TO THE TAB'S NAMESPACE so the two terminals and the rest of the
 * interface agree about where somebody is working. When the tab is on "All
 * namespaces" there IS NO ANSWER — a pod lives in exactly one namespace — so
 * this returns the empty string and the dialog asks.
 *
 * It deliberately does NOT fall back to a system namespace. That is where the
 * node shell's own default points (kube-system, where admission is already
 * permissive and where a node shell has to be), and it is the wrong answer for
 * this: an operator looking at their application's namespace would get a pod
 * in kube-system without being told, in the one namespace they are least
 * likely to be permitted to create one in.
 */
export function clusterShellNamespaceFor(tabNamespace: string): string {
  return tabNamespace === ALL_NAMESPACES ? '' : tabNamespace.trim()
}

/**
 * Normalises the dialog's inputs.
 *
 * A blank image falls back to the default rather than being sent as an empty
 * string the backend would reject. The namespace is NOT given a default: see
 * clusterShellNamespaceFor. A blank one comes back blank, and `canOpen` below
 * is what stops it being sent.
 */
export function clusterShellRequest(image: string, namespace: string): ClusterShellRequest {
  return {
    image: image.trim() || DEFAULT_CLUSTER_SHELL_IMAGE,
    namespace: namespace.trim(),
  }
}

/**
 * Whether the dialog may be confirmed.
 *
 * The namespace is the only thing that can be missing — the image always
 * resolves to a default — and it is what keeps "ask rather than guess" true in
 * the interface as well as in Go.
 */
export function canOpenClusterShell(namespace: string): boolean {
  return namespace.trim() !== ''
}

/**
 * The sentence shown when the tab is on "All namespaces".
 *
 * Says what is missing and why, rather than leaving a disabled button with no
 * explanation beside it.
 */
export const CLUSTER_SHELL_NAMESPACE_PROMPT =
  'This tab is showing every namespace, and a pod lives in exactly one — name the namespace to open the shell in.'

/**
 * The sentence shown beside pods PodSteer created here that are NOT running.
 *
 * They are reported and never offered: an attach to an exited pod fails for a
 * reason the offer gave nobody a way to see. Saying so is what explains a
 * namespace that has been accumulating them.
 */
export function exitedShellsNote(count: number): string {
  if (count <= 0) return ''
  const plural = count === 1 ? 'pod' : 'pods'
  const verb = count === 1 ? 'is' : 'are'
  return `${count} shell ${plural} PodSteer created here ${verb} no longer running, so ${count === 1 ? 'it' : 'they'} cannot be attached to. They are deleted an hour after they were created.`
}
